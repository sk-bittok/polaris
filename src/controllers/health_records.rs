use axum::{
    Json, Router, debug_handler,
    extract::{Path, State},
    http::StatusCode,
    response::{IntoResponse, Response},
    routing::{get, post},
};

use crate::{
    AppContext, Result,
    models::{dto::records::NewHealthRecord, health_records::HealthRecord, users::User},
};

#[debug_handler]
async fn all(user: User, State(ctx): State<AppContext>) -> Result<Response> {
    let models = HealthRecord::find_all(&ctx.db, user.organisation_pid).await?;

    Ok((StatusCode::OK, Json(models)).into_response())
}

#[debug_handler]
async fn add(
    user: User,
    State(ctx): State<AppContext>,
    Json(params): Json<NewHealthRecord<'static>>,
) -> Result<Response> {
    let model = HealthRecord::create(&ctx.db, &params, user.organisation_pid, user.pid).await?;

    Ok((StatusCode::CREATED, Json(model)).into_response())
}

#[debug_handler]
async fn one(user: User, State(ctx): State<AppContext>, Path(id): Path<i32>) -> Result<Response> {
    let model = HealthRecord::find_by_id(&ctx.db, id, user.organisation_pid).await?;

    Ok((StatusCode::OK, Json(model)).into_response())
}

pub fn router(ctx: AppContext) -> Router {
    Router::new()
        .route("/", get(all))
        .route("/", post(add))
        .route("/{id}", get(one))
        .with_state(ctx)
}
