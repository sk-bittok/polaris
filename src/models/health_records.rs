#![allow(clippy::missing_errors_doc)]

use std::str::FromStr;

use chrono::{DateTime, FixedOffset, NaiveDate};
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};
use sqlx::{Encode, Executor, Postgres, postgres::PgQueryResult, prelude::FromRow};
use uuid::Uuid;

use crate::seed::Seedable;

use super::{ModelError, ModelResult, dto::records::NewHealthRecord};

#[derive(Debug, Deserialize, Serialize, FromRow, Encode)]
pub struct HealthRecord {
    pub(crate) id: i32,
    pub(crate) animal_pid: Uuid,
    pub(crate) organisation_pid: Uuid,
    pub(crate) record_type: String,
    pub(crate) record_date: NaiveDate,
    pub(crate) description: String,
    pub(crate) treatment: Option<String>,
    pub(crate) medicine: Option<String>,
    pub(crate) dosage: Option<String>,
    pub(crate) cost: Option<Decimal>,
    pub(crate) performed_by: Option<String>,
    pub(crate) notes: Option<String>,
    pub(crate) created_by: Uuid,
    pub(crate) created_at: DateTime<FixedOffset>,
    pub(crate) updated_at: DateTime<FixedOffset>,
}

impl HealthRecord {
    pub async fn create<'a, C>(
        db: C,
        params: &NewHealthRecord<'_>,
        org_pid: Uuid,
        user_pid: Uuid,
        animal_pid: Uuid,
    ) -> ModelResult<Self>
    where
        C: Executor<'a, Database = Postgres>,
    {
        let record_date = NaiveDate::from_str(&params.date)?;

        let record = sqlx::query_as::<_, Self>(
            "INSERT INTO health_records 
            (animal_pid, organisation_pid, created_by, record_type, record_date,
            description, treatment, medicine, dosage, cost, performed_by, notes)
            VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12) RETURNING * ",
        )
        .bind(animal_pid)
        .bind(org_pid)
        .bind(user_pid)
        .bind(params.record_type.as_ref())
        .bind(record_date)
        .bind(params.description.as_ref())
        .bind(params.treatement.as_ref())
        .bind(params.medicine.as_deref())
        .bind(params.dosage.as_deref())
        .bind(params.cost.map(|cost| Decimal::new(cost, 2)))
        .bind(params.performed_by.as_deref())
        .bind(params.notes.as_deref())
        .fetch_one(db)
        .await?;

        Ok(record)
    }

    pub async fn find_all<'a, C>(db: C, org_pid: Uuid) -> ModelResult<Vec<Self>>
    where
        C: Executor<'a, Database = Postgres>,
    {
        let records =
            sqlx::query_as::<_, Self>("SELECT * FROM health_records WHERE organisation_pid = $1")
                .bind(org_pid)
                .fetch_all(db)
                .await?;

        Ok(records)
    }

    pub async fn find_by_id<'a, C>(db: C, id: i32, org_pid: Uuid) -> ModelResult<Self>
    where
        C: Executor<'a, Database = Postgres>,
    {
        let record = sqlx::query_as::<_, Self>(
            "SELECT * FROM health_records WHERE id = $1 AND organisation_pid = $2",
        )
        .bind(id)
        .bind(org_pid)
        .fetch_optional(db)
        .await?;

        record.ok_or_else(|| ModelError::EntityNotFound)
    }

    pub async fn find_by_record_date_range<'e, C>(
        db: C,
        org_pid: Uuid,
        start_date: NaiveDate,
        end_date: NaiveDate,
    ) -> ModelResult<Vec<Self>>
    where
        C: Executor<'e, Database = Postgres>,
    {
        sqlx::query_as::<_, Self>(
                "SELECT * FROM health_records WHERE organisation_pid = $1 AND record_date BETWEEN $2 AND $3"
            )
            .bind(org_pid)
            .bind(start_date)
            .bind(end_date)
            .fetch_all(db)
            .await
            .map_err(Into::into)
    }

    pub async fn find_by_animal<'e, C>(
        db: C,
        animal_pid: Uuid,
        org_pid: Uuid,
    ) -> ModelResult<Vec<Self>>
    where
        C: Executor<'e, Database = Postgres>,
    {
        sqlx::query_as::<_, Self>(
            "
            SELECT * FROM health_records WHERE animal_pid = $1 AND organisation_pid = $2
            ",
        )
        .bind(animal_pid)
        .bind(org_pid)
        .fetch_all(db)
        .await
        .map_err(Into::into)
    }

    pub async fn find_by_record_type<'e, C>(
        db: C,
        record_type: &str,
        org_pid: Uuid,
    ) -> ModelResult<Vec<Self>>
    where
        C: Executor<'e, Database = Postgres>,
    {
        sqlx::query_as::<_, Self>(
            "SELECT * FROM health_records WHERE organisation_pid = $1 AND record_type LIKE $2",
        )
        .bind(org_pid)
        .bind(format!("%{record_type}%"))
        .fetch_all(db)
        .await
        .map_err(Into::into)
    }

    pub async fn find_by_user<'a, C>(db: C, org_pid: Uuid, user_pid: Uuid) -> ModelResult<Vec<Self>>
    where
        C: Executor<'a, Database = Postgres>,
    {
        sqlx::query_as::<_, Self>(
            "SELECT * FROM health_records WHERE organisation_pid = $1 AND created_by = $2",
        )
        .bind(org_pid)
        .bind(user_pid)
        .fetch_all(db)
        .await
        .map_err(Into::into)
    }

    pub async fn delete_by_id<'e, C>(db: C, org_pid: Uuid, id: i32) -> ModelResult<PgQueryResult>
    where
        C: Executor<'e, Database = Postgres>,
    {
        sqlx::query("DELETE FROM health_records WHERE id = $1 AND organisation_pid = $2")
            .bind(id)
            .bind(org_pid)
            .execute(db)
            .await
            .map_err(Into::into)
    }

    pub async fn seed(db: &sqlx::PgPool, path: &str) -> ModelResult<()> {
        let records = Self::load_file(path).await?;

        Self::seed_data(db, &records).await
    }
}

#[async_trait::async_trait]
impl Seedable for HealthRecord {
    async fn seed_data(db: &sqlx::PgPool, records: &[Self]) -> ModelResult<()> {
        for record in records {
            sqlx::query("
                INSERT INTO health_records (id, animal_pid, organisation_pid, record_type, record_date, description,
                treatment, medicine, dosage, cost, performed_by, notes, created_by, created_at, updated_at) 
                VALUES ( $1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15 )
            ")
                .bind(record.id)
                .bind(record.animal_pid)
                .bind(record.organisation_pid)
                .bind(record.record_type.as_str())
                .bind(record.record_date)
                .bind(record.description.as_str())
                .bind(record.treatment.as_deref())
                .bind(record.medicine.as_deref())
                .bind(record.dosage.as_deref())
                .bind(record.cost)
                .bind(record.performed_by.as_deref())
                .bind(record.notes.as_deref())
                .bind(record.created_by)
                .bind(record.created_at)
                .bind(record.updated_at)
                .execute(db)
                .await?;
        }

        Ok(())
    }
}
