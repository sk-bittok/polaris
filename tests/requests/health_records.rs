use axum::http::StatusCode;
use insta::{Settings, assert_debug_snapshot, with_settings};
use polaris::models::health_records::HealthRecord;
use serial_test::serial;

use crate::requests::prepare_auth;

macro_rules! configure_insta {
    ($(expr:expr),*) => {
        let mut settings = Settings::clone_current();
        settings.set_snapshot_path("snapshots/records");
        settings.set_snapshot_suffix("health");
        settings.set_prepend_module_to_snapshot(false);
        let _guard = settings.bind_to_scope();
    };
}

#[tokio::test]
#[serial]
async fn can_get_all() {
    crate::request(|server, context| async move {
        configure_insta!();

        crate::seed_data(&context.db).await.unwrap();

        let user = prepare_auth::login_user(&server, &context).await;

        let (auth_header, auth_value) = prepare_auth::auth_header(user.access_token);

        let response = server
            .get("/health-records")
            .add_header(auth_header, auth_value)
            .await;

        assert_eq!(response.status_code(), StatusCode::OK);

        assert_debug_snapshot!((response.status_code(), response.text()));
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_get_by_id() {
    crate::request(|server, context| async move {
        configure_insta!();

        crate::seed_data(&context.db).await.unwrap();

        let user = prepare_auth::login_user(&server, &context).await;

        let (auth_header, auth_value) = prepare_auth::auth_header(user.access_token);

        let response = server
            .get("/health-records/101")
            .add_header(auth_header, auth_value)
            .await;

        assert_eq!(response.status_code(), StatusCode::OK);

        assert_debug_snapshot!((response.status_code(), response.text()));
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_post_one() {
    crate::request(|server, context| async move {
        configure_insta!();

        crate::seed_data(&context.db).await.unwrap();

        let user = prepare_auth::login_user(&server, &context).await;
        let (auth_header, auth_value) = prepare_auth::auth_header(user.access_token);

        let body = serde_json::json!({
            "tagId": "AC002",
            "recordDate": "2025-05-12",
            "recordType": "vaccination",
            "description": "Vaccinated Buttercup against foot and mouth disease.",
            "treatment": "oral solution",
            "dosage": "250",
            "medicine": "Anti FnM",
            "cost": 2500,
            "performedBy": "James Dokter",
            "notes": null,
        });

        let response = server
            .post("/health-records")
            .add_header(auth_header, auth_value)
            .json(&body)
            .await;

        assert_eq!(response.status_code(), StatusCode::CREATED);

        with_settings!({
            filters => {
                let mut filters = crate::cleanup_date().to_vec();
                filters.extend(crate::cleanup_uuid().to_vec());
                filters.extend(crate::cleanup_int().to_vec());
                filters
            }
        }, {
        assert_debug_snapshot!((response.status_code(), response.json::<HealthRecord>()));
        })
    })
    .await;
}
