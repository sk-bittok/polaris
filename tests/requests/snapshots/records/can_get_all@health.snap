---
source: tests/requests/health_records.rs
assertion_line: 36
expression: "(response.status_code(), response.text())"
---
(
    200,
    "[{\"id\":101,\"animalPid\":\"b2bd6270-8bec-42ce-99ff-d0eb1a076221\",\"animalName\":\"Daisy\",\"animalTagId\":\"AC001\",\"organisationPid\":\"9d5b0c1e-6a48-4bce-b818-dc8c015fd8a0\",\"organisationName\":\"Acme Corp\",\"recordType\":\"treatment\",\"recordDate\":\"2024-09-08\",\"description\":\"Daisy suffered from East cost fever\",\"treatment\":\"Antibiotics and vitamins administered by injection\",\"medicine\":\"Amoxillin\",\"dosage\":\"450\",\"cost\":\"1650.00\",\"performedBy\":\"John Artz\",\"notes\":null,\"createdBy\":\"bd6f7c26-d2c9-487e-b837-8f77be468033\",\"createdByName\":\"John Doe\",\"createdAt\":\"2024-09-08T11:15:00Z\",\"updatedAt\":\"2024-09-08T11:15:00Z\"},{\"id\":102,\"animalPid\":\"62197c29-a2dd-4591-a4d0-92f6d7099760\",\"animalName\":\"Buttercup\",\"animalTagId\":\"AC002\",\"organisationPid\":\"9d5b0c1e-6a48-4bce-b818-dc8c015fd8a0\",\"organisationName\":\"Acme Corp\",\"recordType\":\"treatment\",\"recordDate\":\"2024-12-12\",\"description\":\"Buttercup suffered from milk fever\",\"treatment\":\"Suppliments and vitamins administered by IV fluid injection\",\"medicine\":\"Calcium rich suppliments\",\"dosage\":\"250\",\"cost\":\"1850.00\",\"performedBy\":\"John Artz\",\"notes\":null,\"createdBy\":\"bd6f7c26-d2c9-487e-b837-8f77be468033\",\"createdByName\":\"John Doe\",\"createdAt\":\"2024-12-12T11:15:00Z\",\"updatedAt\":\"2024-12-12T11:15:00Z\"},{\"id\":103,\"animalPid\":\"d909e761-36da-4062-ae78-abba4f7c1103\",\"animalName\":\"Ferdinand\",\"animalTagId\":\"AC003\",\"organisationPid\":\"9d5b0c1e-6a48-4bce-b818-dc8c015fd8a0\",\"organisationName\":\"Acme Corp\",\"recordType\":\"treatment\",\"recordDate\":\"2024-11-28\",\"description\":\"Ferdinand suffered from East cost fever\",\"treatment\":\"Antibiotics and vitamins administered by injection\",\"medicine\":\"Amoxillin\",\"dosage\":\"650\",\"cost\":\"1750.00\",\"performedBy\":\"John Artz\",\"notes\":null,\"createdBy\":\"bd6f7c26-d2c9-487e-b837-8f77be468033\",\"createdByName\":\"John Doe\",\"createdAt\":\"2024-11-28T11:15:00Z\",\"updatedAt\":\"2024-11-28T11:15:00Z\"}]",
)
